src/averageAges.js

```diff
@@ -1,64 +1,43 @@
 'use strict';
 
-/**
- * Implement calculateMenAverageAge function
- *
- * Function returns average age of men in array. If `century` is specified then
- * function calculates average age only for men who died in this century
- *
- * To calculate century:
- * Divide year of person's death by 100: Math.ceil(person.died / 100)
- *
- * @param {object[]} people
- * @param {number} century - optional
- *
- * @return {number}
- */
+const MALE = 'm';
+const FEMALE = 'f';
+
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const men = people.filter((person) => person.sex === MALE);
+  const filteredMen = century !== undefined
+    ? men.filter((person) => Math.ceil(person.died / 100) === century)
+    : men;
+
+  return averageAge(filteredMen);
 }
 
-/**
- * Implement calculateWomenAverageAge function
- *
- * Function returns average age of women in array. If `withChildren` is
- * specified then function calculates average age only for women with children
- *
- * Hint: To check if a woman has children you should find someone who mention
- * her as mother.
- *
- * @param {object[]} people
- * @param {boolean} withChildren - optional
- *
- * @return {number}
- */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const women = people.filter((person) => person.sex === FEMALE);
+  const filteredWomen = women.filter((woman) => withChildren
+    ? people.some((person) => person.mother === woman.name)
+    : true
+  );
+
+  return averageAge(filteredWomen);
 }
 
-/**
- * Implement calculateAverageAgeDiff function.
- *
- * The function returns an average age difference between a child and his or her
- * mother in the array. (A mother's age at child birth)
- *
- * If `onlyWithSon` is specified then function calculates age difference only
- * for sons and their mothers.
- *
- * @param {object[]} people
- * @param {boolean} onlyWithSon - optional
- *
- * @return {number}
- */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // 1. find a mother of each person (or only for men)
-  // 2. keep people who have mothers in the array
-  // 3. calculate the difference child.born - mother.born
-  // 4. return the average value
+  const filteredChildren = people.filter(person => onlyWithSon
+    ? people.find(mother => mother.name === person.mother
+      && person.sex === MALE)
+    : people.find(mother => mother.name === person.mother)
+  );
+
+  const ageGap = filteredChildren.map(child =>
+    child.born - people.find(mother => mother.name === child.mother).born);
+
+  return ageGap.reduce((acc, age) => acc + age) / ageGap.length;
+}
+
+function averageAge(people) {
```

Comment:
```suggestion
function calculateAverageAge(people) {
```
![image](https://github.com/mate-academy/js_average-ages/assets/96409563/759c7ee7-2e4a-4d1c-bcd3-ab520861d3be)